myia/parser.py

```diff
@@ -0,0 +1,434 @@
+    built-ins and for undefined variable names.
+
+    Lastly, the environment has a mapping from AST nodes to Myia nodes that the
+    parser uses e.g. for mapping binary operators to the correct primitives.
+
+    Functions parsed in different environments will share no nodes in common.
+    Functions parsed in the same environment will use the same Myia objects for
+    the same Python objects.
+
+    """
+
+    def __init__(self, object_map: Dict[Any, ANFNode],
+                 ast_map: Dict[Type[ast.AST], ANFNode]) -> None:
+        """Construct an environment."""
+        self.object_map = object_map
+        self.ast_map = ast_map
+
+    def read(self, varnum: str) -> ANFNode:
+        """Read a variable name from the built-ins namespace."""
+        if hasattr(builtins, varnum):
+            return self.object_map[getattr(builtins, varnum)]
+        else:
+            raise ValueError(varnum)
+
+    def map(self, obj: Any) -> ANFNode:
+        """Map a Python object to an ANF node.
+
+        If the object was already converted, the existing Myia node will be
+        returned.
+
+        """
+        if obj in self.object_map:
+            return self.object_map[obj]
+        if isinstance(obj, (float, int, str)):
+            node = Constant(obj)
+            self.object_map[obj] = node
+            return node
+        if isinstance(obj, FunctionType):
+            parser = Parser(self, obj)
+            graph = parser.parse()
+            node = Constant(graph)
+            self.object_map[obj] = node
+            return node
+        raise ValueError(obj)
+
+
+class Parser:
+    """Parser for a function.
+
+    This class handles the parsing of a single user-defined function. It is
+    responsible for handling e.g. globals.
+
+    """
+
+    def __init__(self, environment: Environment,
+                 function: FunctionType) -> None:
+        """Construct a parser."""
+        self.environment = environment
+        self.function = function
+        self.block_map: Dict[Block, Constant] = {}
+
+    def parse(self) -> Graph:
+        """Parse the function into a Myia graph."""
+        tree = ast.parse(textwrap.dedent(inspect.getsource(self.function)))
+        function_def = tree.body[0]
+        assert isinstance(function_def, ast.FunctionDef)
+        return self._process_function(None, function_def)[1].graph
+
+    def block_op(self, block: 'Block') -> Constant:
+        """Return node representing the function corresponding to a block."""
+        if block in self.block_map:
+            return self.block_map[block]
+        op = Constant(block.graph)
+        self.block_map[block] = op
+        return op
+
+    def read(self, varnum: str) -> ANFNode:
+        """Read a variable from the function's global namespace."""
+        assert isinstance(self.function, FunctionType)
+        if varnum in self.function.__globals__:
+            return self.environment.map(self.function.__globals__[varnum])
```

Comment:
I mentioned this elsewhere, but I think we should be able to set/override the globals dictionary we use here.